cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricRecordsWrapper.java

```diff
@@ -0,0 +1,266 @@
+                                        Constants.Metrics.Tag.PROGRAM_TYPE, Constants.Metrics.Tag.PROGRAM,
+                                        Constants.Metrics.Tag.FLOWLET,
+                                        Constants.Metrics.Tag.INSTANCE_ID)));
+    // app, prg type, prg name, mr task type (for mr task only)
+    rules.add(new Rule(ImmutableList.of(Constants.Metrics.Tag.APP,
+                                        Constants.Metrics.Tag.PROGRAM_TYPE, Constants.Metrics.Tag.PROGRAM,
+                                        Constants.Metrics.Tag.MR_TASK_TYPE)));
+    // app, prg type, prg name, service runnable (for service only)
+    rules.add(new Rule(ImmutableList.of(Constants.Metrics.Tag.APP,
+                                        Constants.Metrics.Tag.PROGRAM_TYPE, Constants.Metrics.Tag.PROGRAM,
+                                        Constants.Metrics.Tag.SERVICE_RUNNABLE)));
+    // component
+    rules.add(new Rule(ImmutableList.of(Constants.Metrics.Tag.COMPONENT)));
+    // component, handler
+    rules.add(new Rule(ImmutableList.of(Constants.Metrics.Tag.COMPONENT, Constants.Metrics.Tag.HANDLER)));
+    // component, handler, method
+    rules.add(new Rule(ImmutableList.of(Constants.Metrics.Tag.COMPONENT, Constants.Metrics.Tag.HANDLER,
+                                        Constants.Metrics.Tag.METHOD)));
+    // component, handler, instance id
+    rules.add(new Rule(ImmutableList.of(Constants.Metrics.Tag.COMPONENT, Constants.Metrics.Tag.HANDLER,
+                                        Constants.Metrics.Tag.INSTANCE_ID)));
+    // component, handler, instance id, tag: stream
+    rules.add(new Rule(ImmutableList.of(Constants.Metrics.Tag.COMPONENT, Constants.Metrics.Tag.HANDLER,
+                                        Constants.Metrics.Tag.INSTANCE_ID),
+                       Constants.Metrics.Tag.STREAM));
+    // dataset name
+    // note: weird rule, but this is what we had before
+    rules.add(new Rule(ImmutableList.of(Constants.Metrics.Tag.DATASET), Constants.Metrics.Tag.DATASET));
+
+    Collections.reverse(rules);
+
+    FANOUT_RULES = ImmutableList.copyOf(rules);
+  }
+
+  private final Iterator<MetricValue> rawIterator;
+  private Iterator<MetricsRecord> current;
+
+  public MetricRecordsWrapper(Iterator<MetricValue> rawIterator) {
+    this.rawIterator = rawIterator;
+  }
+
+
+  @Override
+  public boolean hasNext() {
+    if (current != null && current.hasNext()) {
+      return true;
+    }
+    List<MetricsRecord> fanout = null;
+    while (fanout == null || fanout.isEmpty()) {
+      if (!rawIterator.hasNext()) {
+        return false;
+      }
+
+      fanout = fanout(rawIterator.next());
+    }
+
+    current = fanout.iterator();
+    return true;
+  }
+
+  @Override
+  public MetricsRecord next() {
+    if (!hasNext()) {
+      return null;
+    }
+
+    return current.next();
+  }
+
+  private List<MetricsRecord> fanout(MetricValue metricValue) {
```

Comment:
can you add some comments on what this does.  It's not immediately obvious and takes a while to figure it out.